lib/shelley/src/Cardano/Wallet/Shelley/Network.hs

```diff
@@ -771,6 +761,92 @@ mkTipSyncClient tr np localTxSubmissionQ onTipUpdate onPParamsUpdate onInterpret
         NodeToClientV_3
 
 
+-- Reward Account Balances
+
+-- | Monitors values for keys, and allows clients to @query@ them.
+--
+-- Designed to be used for observing reward balances, where we want to query the
+-- balances of /all/ the wallets' accounts on tip change, and allow wallet
+-- workers to @query@ the value.
+--
+-- One could imagine replacing @query@ with a push-based approach.
+data Observer m key value = Observer
+    { startObserving :: key -> m ()
+    , stopObserving :: key -> m ()
+    , query :: key -> m (Maybe value)
+    }
+
+newRewardBalanceFetcher
+    :: forall sc. Tracer IO (NetworkLayerLog sc)
+    -> W.GenesisParameters
+    -- ^ Used to convert tips for logging
+    -> TQueue IO (LocalStateQueryCmd (CardanoBlock sc) IO)
+    -> IO ( Observer IO W.ChimericAccount W.Coin
+          , Tip (CardanoBlock sc) -> IO ()
+            -- Call on tip-change to refresh
+          )
+newRewardBalanceFetcher tr gp queryRewardQ = do
```

Comment:
It'd be nice to be able to test this bit of logic in isolation. This is currently hard because the management of the keys / accounts is mixed with the logic for fetching the reward. I'd suggest to replace the `ChimericAccount` with a `Ord k => k` and to replace `Coin` with an abstract `a`, and pass a function that yield `a` as argument. So, something like:


 ```hs
data Observer m key value = Observer
    { startObserving :: key -> m ()
    , stopObserving :: key -> m ()
    , query :: key -> m (Maybe value)
    , refresh :: tip -> m tip
    }

newObserver 
    :: Ord key
    => (tip -> [key] -> m (Map key value))
    -> Observer m key value
newObserver getValues = ...
```

Reply:
This way, we can separate the network logic from the observer logic and test them separately (at least, test the observer).

Reply:
> It'd be nice to be able to test this bit of logic in isolation. 

Yes 👍  I alluded to it in https://github.com/input-output-hk/cardano-wallet/pull/2034#discussion_r471493798, but didn't want to explore it before hearing your feedback.